benchmarks/distributed/rpc/ps/benchmark.py

```diff
@@ -0,0 +1,944 @@
+            values.append(metric["elapsed_time"])
+            name = metric["name"]
+        rank_totals[name] = values
+        row = {
+            "name": name,
+            "min": min(values),
+            "max": max(values),
+            "mean": statistics.mean(values),
+            "variance": statistics.variance(values),
+            "stdev": statistics.stdev(values)
+        }
+        df = df.append(row, ignore_index=True)
+    print("metrics for {}={}".format(rank_type, rank))
+    print(tabulate(df, showindex=False, headers=df.columns, tablefmt="grid"))
+    if print_metrics_to_dir:
+        file_name = "data_frames/{}_{}__{}.csv".format(
+            rank_type, rank, file_name
+        )
+        df.to_csv(file_name, encoding='utf-8', index=False)
+    return rank_totals
+
+
+class GradientServer:
+
+    # TODO: cuda sparse kernel
+    """
+        rank - the process number of the gradient server in the world
+
+        gradient_servers - the number of processes running gradient
+        servers in the world
+
+        ddp_trainers - the number of processes running ddp_trainers
+        in the world
+
+        backend - the communication backend selected for distributed
+        training
+
+        use_cuda_rpc - indicates that the RPC should be using Cuda or CPU
+
+        batch_mode - indicates if the gradient server should be waiting to
+        sum all gradients at once to reduce Cuda kernel calls
+    """
+    def __init__(
+        self,
+        rank,
+        gradient_servers,
+        ddp_trainers,
+        use_cuda_rpc,
+        backend,
+        batch_mode
+    ):
+        self.lock = threading.Lock()
+        self.rank = rank
+        self.gradient_servers = gradient_servers
+        self.ddp_trainers = ddp_trainers
+        self.trainer_count = ddp_trainers / gradient_servers
+        remainder = (ddp_trainers % gradient_servers) != 0
+        add = (rank - ddp_trainers + 1) <= (ddp_trainers % gradient_servers)
+        if remainder and add:
+            self.trainer_count += 1
+        self.use_cuda_rpc = use_cuda_rpc
+        self.backend = backend
+        self.batch_mode = batch_mode
+        self.futures = {}
+        self.gradient = {}
+        self.batch_number = 0
+        self.metrics = {}
+
+    def clear_batch_state(self):
+        self.futures.clear()
+        self.gradient.clear()
+
+    def average_gradient_bp_key(self, bp_loc):
+        return "{},{}".format(self.batch_number, bp_loc)
+
+    """
+        gs_rref - the shared pointer to the gradient server assigned
+            to the rank
+
+        rank - the process number in the world
+
+        received_batch_number - ddp_trainer batch number
+
+        backend - the communication backend selected for distributed training
+
+        use_cuda_rpc - indicates that the RPC should be using Cuda or CPU
+
+        batch_mode - indicates if the gradient server should
+            sum all gradients at once to reduce Cuda kernel calls
```

Comment:
Curious, what's the motivation of putting everything into a single `GradientServer` class instead of a modularized solution by, say, having multiple (sub-) classes of gradient servers?

One problem of the current implementation is that this single class contains a lot of branches, making it less readable and hard to tell where we should modify if we would like to add some test cases. 

Another downside is that the docs cannot properly match the arguments of the functions here, and we have to use a `kwargs` to manually retrieve optional args based on the configuration.

Reply:
I didn't have extensibility in mind when writing the gradient server class. I will make changes to make it modular.